src/shared/utilities/timeoutUtils.ts

```diff
@@ -0,0 +1,50 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+/**
+ * Timeout coTimeoutbject that can handle both cancellation token- and time limit-style timeout situations.
+ * @param timeoutLength Length of timeout duration (in ms)
+ */
+export class Timeout {
+    private readonly _startTime: number
+    private readonly _endTime: number
+    private readonly _timer: Promise<boolean>
+    public constructor(timeoutLength: number) {
+        this._startTime = new Date().getTime()
+        this._endTime = this._startTime + timeoutLength
+        this._timer = new Promise<boolean>((resolve) => {
+            setTimeout(() => resolve(false), timeoutLength)
+        })
+    }
+
+    /**
+     * Returns the amount of time left from the initialization of time Timeout object and with the timeoutLength
+     * Bottoms out at 0
+     */
+    public get remainingTime(): number {
+        const curr = new Date().getTime()
+
+        return (this._endTime - curr > 0 ? this._endTime - curr : 0)
+    }
+
+    /**
+     * Returns a promise that times out after timeoutLength ms have passed since Timeout object initialization
+     * Use this in Promise.race() calls in order to time out awaited functions
+     * Once this timer has finished, cannot be restarted
+     */
+    public get timer(): Promise<boolean> {
```

Comment:
Is there significance to the Promise value or is `Promise<void>` more appropriate?

Reply:
Addressed below, this was used in order to detect the closure of the promise. Will use void for now (with the function running code against the timeout responsible for tracking the state of this promise) and will look into implementing a timed promise object.

Reply:
✔️ 
Caveat though--returning void from a promise will trigger any promise rejection handlers. We need to specifically await this as a value in order to pull an `undefined` without triggering an error.